### PR TITLE
8340453: C2: Improve encoding of LoadNKlass for compact headers

### PR DESCRIPTION
**Pull request by rkennke**
We currently use the offset 4 as a placeholder in LoadNKlass, when running with compact headers. In reality, we are loading from offset 0, but we want to keep LoadNKlass on a separate memory slice from other mark-word-accesses, because LoadNKlass is essentially immutable memory. The consequence is that we need to figure out the address of the mark-word in the backend, and this is ugly.

However, we can do better. We can just as well load 4 bytes from offset 4, and shift by a 32 smaller shift. This has previously not been possible because we needed to check for the monitor bit in the markWord, but this is no longer necessary. This simplifies the code and even makes the instructions encoding a bit smaller.

Testing:
 - [x] tier1 aarch64 +UCOH 
 - [x] tier1 x86_64 +UCOH

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340453](https://bugs.openjdk.org/browse/JDK-8340453): C2: Improve encoding of LoadNKlass for compact headers (**Enhancement** - P4)

### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22078/head:pull/22078` \
`$ git checkout pull/22078`

Update a local copy of the PR: \
`$ git checkout pull/22078` \
`$ git pull https://git.openjdk.org/jdk.git pull/22078/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22078`

View PR using the GUI difftool: \
`$ git pr show -t 22078`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22078.diff">https://git.openjdk.org/jdk/pull/22078.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22078#issuecomment-2474449671)
</details>

### CODE REVIEW

**Comment by dean-long** [comment]
It looks like this only works for little-endian.  Is that documented somewhere?

**Comment by rkennke** [comment]
> It looks like this only works for little-endian. Is that documented somewhere?

I am not sure what you mean. This change is about x86_64 and aarch64, and both are little-endian. The layout of the mark-word is documented in markWord.hpp. Is that what you are looking for?

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 1]
```diff
@@ -6695,16 +6691,21 @@ instruct loadNKlass(iRegNNoSp dst, memory4 mem)
...
+instruct loadNKlassCompactHeaders(iRegNNoSp dst, memory4 mem)
 %{
   match(Set dst (LoadNKlass mem));
   predicate(!needs_acquiring_load(n) && UseCompactObjectHeaders);
 
   ins_cost(4 * INSN_COST);
-  format %{ "load_narrow_klass_compact  $dst, $mem\t# compressed class ptr" %}
+  format %{
+    "ldrw  $dst, $mem\t# compressed class ptr, shifted"
+    "lsrw  $dst, markWord::klass_shift_at_offset"
```
Comment (robcasloz) [inline src/hotspot/cpu/aarch64/aarch64.ad:6702]:
```suggestion
    "lsrw  $dst, $dst, markWord::klass_shift_at_offset"
```

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 2]
```diff
@@ -6695,16 +6691,21 @@ instruct loadNKlass(iRegNNoSp dst, memory4 mem)
...
-instruct loadNKlassCompactHeaders(iRegNNoSp dst, memory_noindex mem)
+instruct loadNKlassCompactHeaders(iRegNNoSp dst, memory4 mem)
 %{
   match(Set dst (LoadNKlass mem));
   predicate(!needs_acquiring_load(n) && UseCompactObjectHeaders);
 
   ins_cost(4 * INSN_COST);
-  format %{ "load_narrow_klass_compact  $dst, $mem\t# compressed class ptr" %}
+  format %{
+    "ldrw  $dst, $mem\t# compressed class ptr, shifted"
```
Comment (robcasloz) [inline src/hotspot/cpu/aarch64/aarch64.ad:6701]:
```suggestion
    "ldrw  $dst, $mem\t# compressed class ptr, shifted\n\t"
```

File: src/hotspot/cpu/x86/x86_64.ad [hunk 3]
```diff
@@ -4368,11 +4368,15 @@ instruct loadNKlassCompactHeaders(rRegN dst, memory mem, rFlagsReg cr)
   match(Set dst (LoadNKlass mem));
   effect(KILL cr);
   ins_cost(125); // XXX
-  format %{ "load_narrow_klass_compact    $dst, $mem\t# compressed klass ptr" %}
+  format %{
+    "movl    $dst, $mem\t# compressed klass ptr, shifted"
```
Comment (robcasloz) [inline src/hotspot/cpu/x86/x86_64.ad:4372]:
```suggestion
    "movl    $dst, $mem\t# compressed klass ptr, shifted\n\t"
```

**Comment by rkennke** [comment]
Thanks, @robcasloz ! I made the suggested changes.
Yeah, I agree, it is still not ideal. But it seems the most-correct way to handle it. Re-shaping LoadNKlass as you suggested would break the semantics of a LoadNode and I am not sure about the subtle or not-so-subtle consequences of that. It seems most-correct to have LoadNKlass use offset 0, but that also sounds scary and potentially affecting all mark-word accesses. Actually using the offset 4 currently seems the sanest approach to me.

**Review by robcasloz — APPROVED** [review]
Thanks for applying the suggestions. I agree that there is no obvious better solution, at least while both original and compact object headers have to coexist.

**Comment by dean-long** [comment]
> > It looks like this only works for little-endian. Is that documented somewhere?
> 
> I am not sure what you mean. This change is about x86_64 and aarch64, and both are little-endian. The layout of the mark-word is documented in markWord.hpp. Is that what you are looking for?

Wouldn't klass_offset_in_bytes need to have a different value for big-endian?

**Comment by rkennke** [comment]
> > > It looks like this only works for little-endian. Is that documented somewhere?
> > 
> > 
> > I am not sure what you mean. This change is about x86_64 and aarch64, and both are little-endian. The layout of the mark-word is documented in markWord.hpp. Is that what you are looking for?
> 
> Wouldn't klass_offset_in_bytes need to have a different value for big-endian?

I think so, yes. It would be 0, right? Which would not easily work, because offset 0 would conflict with the C2 memory slice for the mark-word. Big endian arches (do we even support any?) would need to use the old fake-offset method like we did before this change.

**Comment by dean-long** [comment]
Yes, I think we still care about big-endian.  See for example JDK-8314949 and JDK-8312495.

**Comment by TheRealMDoerr** [comment]
AIX on PPC64 and linux on s390 are supported Big Endian platforms. We can't do the same change for these platforms without endianness sensitive adaptations.

**Pull request closed without merging**